### PR TITLE
418 - running pre-caching for the attributes cardinality during startup

### LINKED ISSUE #418: Run pre-caching for the attributes cardinality during startup
Currently it's being triggered via endpoint.

### PR DESCRIPTION
**Pull request by piotrkosecki**
resolves #418

### CODE REVIEW

File: src/main/scala/tech/cryptonomic/conseil/tezos/TezosPlatformDiscoveryOperations.scala [hunk 1]
```diff
@@ -295,6 +297,7 @@ class TezosPlatformDiscoveryOperations(
             } else {
               (for {
                 _ <- caching.putAttributes(tableName, attributes)
+                _ <- IO.contextShift(executionContext).shift
```
Comment (ba205) [inline src/main/scala/tech/cryptonomic/conseil/tezos/TezosPlatformDiscoveryOperations.scala]: Should switching to IO be put in this pr, or did it directly help with the caching task?

Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/conseil/tezos/TezosPlatformDiscoveryOperations.scala]: that was a change requested by @ivanopagano in previous caching PR, but it was merged before I could push it, so I thought I'll add it here

File: src/main/scala/tech/cryptonomic/conseil/tezos/TezosPlatformDiscoveryOperations.scala [hunk 2]
```diff
@@ -177,6 +177,7 @@ class TezosPlatformDiscoveryOperations(
         } else {
           (for {
             _ <- caching.putEntities(networkName, ent)
+            _ <- IO.contextShift(executionContext).shift
```
Comment (ivanopagano) [inline src/main/scala/tech/cryptonomic/conseil/tezos/TezosPlatformDiscoveryOperations.scala]: If it helps, you can reduce repetition by making an implicit ContextShift[IO] available at the class beginning, and simply call `IO.shift` henceforth

File: src/main/scala/tech/cryptonomic/conseil/tezos/TezosPlatformDiscoveryOperations.scala [hunk 3]
```diff
@@ -360,40 +363,29 @@ class TezosPlatformDiscoveryOperations(
...
+     }
+
+    result.flatMap(_.toList.sequence) >> caching.updateCachingStatus(Finished)
+  }.unsafeToFuture()
 
   /** Helper method for updating */
   private def getAllUpdatedAttributes(entities: List[Entity], attributes: Cache[List[Attribute]]): IO[Map[String, List[Attribute]]] = {
     val queries = attributes
-      .filterKeys(entities.map(_.name).toSet)
+      .filterKeys {
```
Comment (ivanopagano) [inline src/main/scala/tech/cryptonomic/conseil/tezos/TezosPlatformDiscoveryOperations.scala]: this was quite subtle

File: src/main/scala/tech/cryptonomic/conseil/Conseil.scala [hunk 4]
```diff
@@ -53,6 +53,11 @@ object Conseil extends App with LazyLogging with EnableCORSDirectives with Conse
         case Success(_) => logger.info("Pre-caching successful!")
       }
 
+      tezosPlatformDiscoveryOperations.initAttributesCache.onComplete {
```
Comment (ivanopagano) [inline src/main/scala/tech/cryptonomic/conseil/Conseil.scala]: Did we consider doing retries? Maybe in a separate PR?

File: src/main/scala/tech/cryptonomic/conseil/tezos/TezosPlatformDiscoveryOperations.scala (hunk 3 above)
Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/conseil/tezos/TezosPlatformDiscoveryOperations.scala]: yes 😄

File: src/main/scala/tech/cryptonomic/conseil/Conseil.scala (hunk 4 above)
Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/conseil/Conseil.scala]: I think it should be don in other PR for this and for preCaching

File: src/main/scala/tech/cryptonomic/conseil/tezos/TezosPlatformDiscoveryOperations.scala (hunk 2 above)
Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/conseil/tezos/TezosPlatformDiscoveryOperations.scala]: ok, done

**Comment by dorians** [comment]
Looks good, my only concern is what happens when we run `Lorre`? Should we rerun `Conseil` after that to have cardinality up to date?

**Pull request merged**